flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -129,6 +129,48 @@ convertOptExtentExpr(Fortran::lower::AbstractConverter &converter,
   return fir::getBase(converter.genExprValue(&e, stmtCtx, loc));
 }
 
+/// Return the extended value for a component of a derived type instance given
+/// the extended value \p obj of the derived type instance and the address of
+/// the component.
+static fir::ExtendedValue
+componentToExtendedValue(Fortran::lower::FirOpBuilder &builder,
+                         mlir::Location loc, const fir::ExtendedValue &obj,
+                         mlir::Value component) {
+  auto fieldTy = fir::dyn_cast_ptrEleTy(component.getType());
+  if (fieldTy.dyn_cast<fir::BoxType>())
+    fir::emitFatalError(loc, "TODO: lower pointer/allocatable component ref");
```

Comment:
I think using the TODO macro will make this more consistent, easier to read.

Reply:
OK, I agree that is more consistent, I started a bit using fatal error for TODOs so that we can see which line in the Fortran program is hitting the TODO, I think it makes @sscalpone life easier. Could we add a TODO macro that can take an mlir::Location  and uses a fir::emitFatalError under the hood ?
Something like `TODO_AT(loc, "lower pointer/allocatable component ref");`
I'll go for regular TODO for now and make TODO_AT in its own PR if that seems like a good idea.

Reply:
I disagree that the _source_ line has any relevance to an unimplemented feature. TODOs are orthogonal to error messages and serve a different purpose.

If we absolutely have to have source lines emitted mixed in with information that points developers to the correct place, then we should update `TODO` to take a second `Location` argument. You'll want a second macro that provides a dummy unknown location since there will be places where a source position is illogical.